system/lib/wasmfs/wasmfs.cpp

```diff
@@ -108,8 +108,13 @@ __wasi_errno_t __wasi_fd_close(__wasi_fd_t fd) {
 }
 
 long __syscall_fstat64(long fd, long buf) {
-  // Release FileTable lock after accessing desired open file.
-  auto file = FileTable::get()[fd].locked().getFile();
+  auto openFile = FileTable::get()[fd];
+
+  if (!openFile) {
+    return -(ENOENT);
```

Comment:
It looks like `EBADF` doesn't work because musl falls back to trying `stat` or `fstatat` whne `fstat` returns EBADF but the file descriptor is valid:

https://github.com/emscripten-core/emscripten/blob/main/system/lib/libc/musl/src/stat/fstat.c#L23-L28

@sbc100 confirmed that it would be reasonable to skip those fallbacks when `__EMSCRIPTEN__` is defined, which would allow returning EBADF here to work correctly.